text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,412 @@
+dependencies directly into the source directory at well-known locations,
+though this would mean that if the source of those dependencies were to
+change, the user would have to re-run the build system (rather than just
+run `cargo`) to refresh the vendored dependency. And this approach too
+would end up polluting the user's source directory.
+
+An earlier iteration of the design avoided adding a new field to
+dependencies, and instead inlined the base name into the path using
+`path = "base::relative/path"`. This has the advantage of not
+introducing another special keyword in `Cargo.toml`, but comes at the
+cost of making `::` illegal in paths, which was deemed too great.
+
+Alternatively, we could add support for extrapolating environment
+variables (or arbitrary configuration values?) in `Cargo.toml` values.
+That way, the path could be given as `path =
+"${base.name}/relative/path"`. While that works, it's not trivially
+backwards compatible, may be confusing when users try to extrapolate
+random other configuration variables in their paths, and _seems_ like a
+possible Pandora's box of corner-cases.
+
+The [`[paths]`
+feature](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#paths-overrides)
+could be updated to lift its current limitations around adding
+dependencies and requiring that the dependencies be available on
+crates.io. This would allow users to avoid `path` dependencies in more
+cases, but makes the replacement more implicit than explicit. That
+change is also more likely to break existing users, and to involve
+significant refactoring of the existing mechanism.
+
+We could add another type of local registry that is explicitly declared
+in `Cargo.toml`, and from which local dependencies could then be drawn.
+Something like:
+
+```toml
+[registry.local]
+path = "/path/to/path/registry"
+```
+
+This would make specifying the dependencies somewhat nicer (`version =
+"1", registry = "local"`), and would ensure a standard layout for the
+locations of the local dependencies. However, using local dependencies
+in this manner would require more set-up to arrange for the right
+registry layout, and we would be introducing what is effectively a
+mutable registry, which Cargo has avoided thus far.
+
+Even with such an approach, there are benefits to being able to not put
+complex paths into `Cargo.toml` as they may differ on other build hosts.
+So, a mechanism for indirecting through a path name may still be
+desirable.
+
+Ultimately, by not having a mechanism to name paths that lives outside
+of `Cargo.toml`, we are forcing developers to coordinate their file
+system layouts without giving them a mechanism for doing so. Or to work
+around the lack of a mechanism by requiring developers to add symlinks
+in strategic locations, cluttering their directories. The proposed
+mechanism is simple to understand and to use, and still covers a wide
+variety of use-cases.
+

```

Comment:
```suggestion

## Search path vs workspace inheritance

This RFC proposes using a search order for path bases: Cargo config (`config.toml` or environment), then the manifest, then the workspace manifest if any, then Cargo's built-in path bases. Rather than including the workspace manifest in the search order, we could instead use workspace inheritance. This RFC proposes including the workspace in the search order, for multiple reasons:

This is something that crates in a workspace will almost always want to inherit. Having to add
```toml
[base-paths]
workspace = true
```
adds noise to what is supposed to be a simplifying/shortening feature. This would be even more complex and verbose if overriding a single `base-path` but using others: adding such an override would require listing all inherited paths individually except for the one being overridden.

Unlike some other data in the manifest, `base-paths` are *only* used if explicitly referenced by name. Thus, there's no danger of "invisible" inheritance of semantically meaningful fields. If you implicitly inherit `package.edition` with no visible sign in your manifest, that could create confusion. If you implicitly inherit `base-paths`, there will still be a visible sign in your manifest: using `base = "name"` in a dependency. And if you want a different value for that base, you can override it in your manifest's `base-paths`.

The use of search order to obtain `base-paths` from `config.toml`, the manifest, and cargo's built-in bases, in that order of priority, gives the right semantics for overriding/shadowing `base-paths`.  Given that model, inserting the workspace base-paths into that model seems like a natural fit. Using workspace inheritance for base-paths would be more complicated, combining two different mechanisms rather than just one ordered search.
```

Reply:
For design discussion on whether we should go with this, see https://github.com/rust-lang/rfcs/pull/3529#discussion_r1649339907

Reply:
Removed support for specifying path bases in manifests, so this is moot.